### PR TITLE
CropworkspaceRagged

### PR DESCRIPTION
**Pull request by peterfpeterson**
This is part of a workflow for total scattering that is described in the `.rst` file.

**To test:**

It is easiest to try out with the additional test file and various cropping parameters.

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        numMax = len(xmaxs)
+
+        errors = {}
+
+        if numMin == 0 and numMax == 0:
+            errors['XMin'] = 'Must specify minimums or maximums'
+            errors['XMax'] = 'Must specify minimums or maximums'
+
+        if numMin > 1 and numMin != numSpec:
+            errors['XMin'] = 'Must specify min for each spectra ({}!={})'.format(numMin, numSpec)
```
Comment (soininen) [inline Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py]: `XMin` or `Xmin`?

File: Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        if numMax > 1 and numMax != numSpec:
+            errors['XMax'] = 'Must specify max for each spectra ({}!={})'.format(numMax, numSpec)
```
Comment (soininen) [inline Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py]: `XMax` or `Xmax`?

File: Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                # accumulate
+                if accumulationWS is None:
+                    accumulationWS = name # messes up progress during very first step
+                else:
+                    ConjoinWorkspaces(InputWorkspace1=accumulationWS,
+                                      InputWorkspace2=name,
+                                      startProgress=(progStart + progStep), endProgress=(progStart + 2 * progStep),
+                                      EnableLogging=False)
+
+            RenameWorkspace(InputWorkspace=accumulationWS, OutputWorkspace=outputWS, EnableLogging=False)
```
Comment (soininen) [inline Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py]: With this line, when running the algorithm with `StoreInADS=False` the output workspace will still show up in the ADS. One solution would be to replace this line by
```python
outputWS = mtd[accumulationWS]
mtd.remove(accumulationWS)
```

and at the end of `PyExec()`:
```python
self.setProperty('OutputWorkspace', outputWS)
```

File: Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    def summary(self):
+        return 'Crop each spectrum of a workspace independently'
+
+    def PyInit(self):
+        self.declareProperty(WorkspaceProperty('InputWorkspace', '',
+                                               direction=Direction.Input), 'input workspace')
+        self.declareProperty(WorkspaceProperty('OutputWorkspace', '',
+                                               direction=Direction.Output), 'output workspace')
+        self.declareProperty(FloatArrayProperty('Xmin'), 'minimum x values with NaN meaning no minimum')
+        self.declareProperty(FloatArrayProperty('Xmax'), 'maximum x values with NaN meaning no maximum')
```
Comment (soininen) [inline Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py]: Is it necessary to accept NaNs here? Wouldn't `-np.inf` and `np.inf` do the job as well and be more logical?

File: Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        return errors
+
+    def PyExec(self):
+        inputWS = self.getProperty('InputWorkspace').value
+        outputWS = self.getProperty('OutputWorkspace').valueAsStr
+        xmins = self.getProperty('XMin').value
+        xmaxs = self.getProperty('XMax').value
+
+        if len(xmins) == 1 and len(xmaxs) == 1:
+            CropWorkspace(InputWorkspace=inputWS, OutputWorkspace=outputWS,
```
Comment (soininen) [inline Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py]: With this line, when running the algorithm with `StoreInADS=False` the output workspace will still show up in the ADS. One solution would be to replace this line by
```python
name = '__{}_cropped_'.format(outputWS)
CropWorkspace(InputWorkspace=inputWS, OutputWorkspace=name, XMin=xmins[0], XMax=xmaxs[0])
outputWS = mtd[name]
mtd.remove(name)
```

and at the end of PyExec():
```python
self.setProperty('OutputWorkspace', outputWS)
```

**Review by soininen — CHANGES_REQUESTED** [review]
The algorithm works mostly as expected, except the following script crashes due to some `blocksize()` problems:
```python
import numpy

xmins = numpy.full(200, 2500.)
xmins[11] = 3100.
xmaxs = numpy.full(200, 5500.)
xmaxs[12] = 4700.
ws = CreateSampleWorkspace()
cropped = CropWorkspaceRagged(ws, Xmin=xmins, Xmax=xmaxs)
print(cropped.getNumberHistograms())
```

Also, the following script seems to pad the 11th histogram with zeros. Is this expected behavior and if so, should it be documented?
```python
import numpy

xmins = numpy.full(200, 2500.)
xmins[11] = 3300.
xmaxs = numpy.full(200, 5500.)
xmaxs[11] = 4700.
ws = CreateSampleWorkspace()
cropped = CropWorkspaceRagged(ws, Xmin=xmins, Xmax=xmaxs)
```

File: Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py (hunk 3 above)
Comment (peterfpeterson) [reply Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py]: This one created an invalid pointer.

**Comment by peterfpeterson** [comment]
I added your first snippet as an additional test. Is the second one meant to be identical?

File: Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py (hunk 3 above)
Comment (soininen) [reply Framework/PythonInterface/plugins/algorithms/CropWorkspaceRagged.py]: Could you check if the attached diff would solve the StoreInADS problem? It also removes a few commented-out lines of leftover code. The file is not compressed, just remove the `.zip` extension.

[CropWorkspaceRagged_fix_StoreInADS.diff.zip](https://github.com/mantidproject/mantid/files/1625489/CropWorkspaceRagged_fix_StoreInADS.diff.zip)

**Comment by peterfpeterson** [comment]
The windows build is unstable because another process is using a temp file that this build is associated with. The doxygen job succeeded, but didn't update the status on github.

**Review by soininen — APPROVED** [review]

**Pull request merged**